keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1087 @@
+available on more than just one node. However, for technical reasons
+(high-speed network for data transfer inside a single data center) or
+political reasons (data must only be stored and processed in a single
+jurisdication) availability may still be limited to a subset of the
+nodes in a cluster.
+
+#### Custom schedulers
+
+For situations not handled by the Kubernetes scheduler now and/or in
+the future, a [scheduler
+extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md)
+can influence pod scheduling based on the information exposed via the
+new API. The
+[topolvm-scheduler](https://github.com/cybozu-go/topolvm/blob/master/docs/design.md#how-the-scheduler-extension-works)
+currently does that with a driver-specific way of storing capacity
+information.
+
+Alternatively, the [scheduling
+framework](https://kubernetes.io/docs/concepts/configuration/scheduling-framework/)
+can be used to build and run a custom scheduler where the desired
+policy is compiled into the scheduler binary.
+
+#### Operators for applications
+
+Application operators for modern scale out storage services
+(e.g. MongoDB, ElasticSearch, Kafka, MySQL, PostgreSQL, Minio, etc.)
+may want to control creation of volumes carefully in order to optimize
+availability, durability, performance and cost. For more information
+about this, see the [StoragePool API for Advanced Storage Placement
+KEP](https://github.com/kubernetes/enhancements/pull/1347).
+
+The storage pool API as introduced in this KEP enables the creation of such
+operators. When combined with a CSI extension that allows a sidecar to
+list and query storage pools, even more information might become
+available, like health of a storage pool.
+
+## Proposal
+
+### Caching remaining capacity via the API server
+
+CSI defines the `GetCapacity` RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new type `CSIStoragePool` type gets
+added to store the result of the `GetCapacity` calls for one storage
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+The external-provisioner will be extended to handle the management
+of `CSIStoragePool` objects. For local storage in the node, there will
+be one such `CSIStoragePool` object per node and CSI driver on that
+node. For network attached storage, there will be one `CSIStoragePool`
+per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver. However, how the external-provisioner discovers these pools
+is currently open.
+
+### Size of ephemeral inline volumes
+
+CSI drivers that allocate storage from kubelet's root directory are
```

Comment:
I don't think this is accurate. The local ephemeral storage feature only counts capacity used by container writeable layer, logs and emptydir volumes.